src/utilcode/util.cpp

```diff
@@ -630,55 +631,88 @@ BYTE * ClrVirtualAllocWithinRange(const BYTE *pMinAddr,
         return NULL;
     }
 
-    // We will do one scan: [pMinAddr .. pMaxAddr]
-    // Align to 64k. See docs for VirtualAllocEx and lpAddress and 64k alignment for reasons.
-    BYTE *tryAddr = (BYTE *)ALIGN_UP((BYTE *)pMinAddr, VIRTUAL_ALLOC_RESERVE_GRANULARITY);
+    // We will do one scan from [pMinAddr .. pMaxAddr]
+    // First align the tryAddr up to next 64k base address. 
+    // See docs for VirtualAllocEx and lpAddress and 64k alignment for reasons.
+    //
+    BYTE *   tryAddr            = (BYTE *)ALIGN_UP((BYTE *)pMinAddr, VIRTUAL_ALLOC_RESERVE_GRANULARITY);
+    bool     success            = false;
+    bool     virtualQueryFailed = false;
+    bool     faultInjected      = false;
+    unsigned virtualQueryCount  = 0;
 
     // Now scan memory and try to find a free block of the size requested.
     while ((tryAddr + dwSize) <= (BYTE *) pMaxAddr)
     {
         MEMORY_BASIC_INFORMATION mbInfo;
-
+            
         // Use VirtualQuery to find out if this address is MEM_FREE
         //
+        virtualQueryCount++;
         if (!ClrVirtualQuery((LPCVOID)tryAddr, &mbInfo, sizeof(mbInfo)))
+        {
+            // Exit and return nullptr if the VirtualQuery call fails.
+            virtualQueryFailed = true;
             break;
-
+        }
+            
         // Is there enough memory free from this start location?
-        // The PAL version of VirtualQuery sets RegionSize to 0 for free
-        // memory regions, in which case we go just ahead and try
-        // VirtualAlloc without checking the size, and see if it succeeds.
-        if (mbInfo.State == MEM_FREE &&
-            (mbInfo.RegionSize >= (SIZE_T) dwSize || mbInfo.RegionSize == 0))
+        if ((mbInfo.State == MEM_FREE) && (mbInfo.RegionSize >= (SIZE_T)dwSize))
```

Comment:
Yes, with the new check we always call VirtualAlloc, without knowing if the size is large enough.

I'm planning on adding a couple of comments to highlight what happens here on UNIX:

```
        // Is there enough memory free from this start location?
        // Note that for UNIX the mbInfo.RegionSize returned will alwys be 0
        if ((mbInfo.State == MEM_FREE) && 
            (mbInfo.RegionSize >= (SIZE_T) dwSize || mdInfo.RegionSize == 0)) 
        {
            // Try reserving the memory using VirtualAlloc now
            pResult = (BYTE*)ClrVirtualAlloc(tryAddr, dwSize, MEM_RESERVE, flProtect);

            // Normally this will be successful
            //
            if (pResult != nullptr)
            {
                success = true;
                // return pResult
                break;
            }

#ifdef _DEBUG
            if (ShouldInjectFaultInRange())
            {
                // return nullptr (failure)
                faultInjected = true;
                break;
            }
#endif // _DEBUG

            // On UNIX we will also fail if our request size 'dwSize' is larger than 64K and
            // and our tryAddr is pointing at a small MEM_FREE region (smaller than 'dwSize')
            // However we can't distinguish between this and the race case.

            // We might fail in a race.  So just move on to next region and continue trying
            tryAddr = tryAddr + VIRTUAL_ALLOC_RESERVE_GRANULARITY;
        }
```